xds/internal/balancer/ringhash/ringhash.go

```diff
@@ -208,55 +219,51 @@ type ringhashBalancer struct {
 // SubConn states are Idle.
 func (b *ringhashBalancer) updateAddresses(addrs []resolver.Address) bool {
 	var addrsUpdated bool
-	// addrsSet is the set converted from addrs, it's used for quick lookup of
-	// an address.
-	//
-	// Addresses in this map all have attributes stripped, but metadata set to
-	// the weight. So that weight change can be detected.
-	//
-	// TODO: this won't be necessary if there are ways to compare address
-	// attributes.
-	addrsSet := make(map[resolver.Address]struct{})
-	for _, a := range addrs {
-		aNoAttrs := a
-		// Strip attributes but set Metadata to the weight.
-		aNoAttrs.Attributes = nil
-		w := weightedroundrobin.GetAddrInfo(a).Weight
-		if w == 0 {
+	// addrsSet is the set converted from addrs, it's used for quick lookup of an
+	// address. Key type here is the same as that of the `subConns` map.
+	addrsSet := resolver.NewAddressMap()
+	for _, addr := range addrs {
+		addrInfo := weightedroundrobin.GetAddrInfo(addr)
+		if addrInfo.Weight == 0 {
+
 			// If weight is not set, use 1.
-			w = 1
+			addrInfo.Weight = 1
 		}
-		aNoAttrs.Metadata = w
-		addrsSet[aNoAttrs] = struct{}{}
-		if scInfo, ok := b.subConns[aNoAttrs]; !ok {
+		modifiedAddr := addr
+		modifiedAddr.Attributes = nil
```

Comment:
Why clear `Attributes`?  That seems wrong, since `Attributes` will affect subchannel creation.  I think we should be clearing `BalancerAttributes` if anything (or just leave it alone since it doesn't matter for comparisons?).

Reply:
If you see line 240, we use the original `addr` for subchannel creation, thereby passing all the attributes that we received. This `modifiedAddr` is used only to key the `subConns` map and here we are interested only in the address weight, which is currently stored as part of the `BalancerAttributes` field.

Reply:
I have changed this to not ignore `Attributes`.